rr_graph_uxsdcxx.cpp

```diff
@@ -0,0 +1,3174 @@
+	return enum_switch_type::UXSD_INVALID;
+}
+
+inline enum_pin_type lex_pin_type(const char *in, bool throw_on_invalid){
+	unsigned int len = strlen(in);
+	switch(len){
+	case 4:
+		switch(*((triehash_uu32*)&in[0])){
+		case onechar('O', 0, 32) | onechar('P', 8, 32) | onechar('E', 16, 32) | onechar('N', 24, 32):
+			return enum_pin_type::OPEN;
+		break;
+		default: break;
+		}
+		break;
+	case 5:
+		switch(*((triehash_uu32*)&in[0])){
+		case onechar('I', 0, 32) | onechar('N', 8, 32) | onechar('P', 16, 32) | onechar('U', 24, 32):
+			switch(in[4]){
+			case onechar('T', 0, 8):
+				return enum_pin_type::INPUT;
+			break;
+			default: break;
+			}
+		break;
+		default: break;
+		}
+		break;
+	case 6:
+		switch(*((triehash_uu32*)&in[0])){
+		case onechar('O', 0, 32) | onechar('U', 8, 32) | onechar('T', 16, 32) | onechar('P', 24, 32):
+			switch(in[4]){
+			case onechar('U', 0, 8):
+				switch(in[5]){
+				case onechar('T', 0, 8):
+					return enum_pin_type::OUTPUT;
+				break;
+				default: break;
+				}
+			break;
+			default: break;
+			}
+		break;
+		default: break;
+		}
+		break;
+	default: break;
+	}
+	if(throw_on_invalid)
+		throw std::runtime_error("Found unrecognized enum value " + std::string(in) + " of enum_pin_type.");
+	return enum_pin_type::UXSD_INVALID;
+}
+
+inline enum_loc_side lex_loc_side(const char *in, bool throw_on_invalid){
+	unsigned int len = strlen(in);
+	switch(len){
+	case 3:
+		switch(in[0]){
+		case onechar('T', 0, 8):
+			switch(in[1]){
+			case onechar('O', 0, 8):
+				switch(in[2]){
+				case onechar('P', 0, 8):
+					return enum_loc_side::TOP;
+				break;
+				default: break;
+				}
+			break;
+			default: break;
+			}
+		break;
+		default: break;
+		}
+		break;
+	case 4:
+		switch(*((triehash_uu32*)&in[0])){
+		case onechar('L', 0, 32) | onechar('E', 8, 32) | onechar('F', 16, 32) | onechar('T', 24, 32):
+			return enum_loc_side::LEFT;
+		break;
+		default: break;
+		}
+		break;
+	case 5:
+		switch(*((triehash_uu32*)&in[0])){
+		case onechar('R', 0, 32) | onechar('I', 8, 32) | onechar('G', 16, 32) | onechar('H', 24, 32):
+			switch(in[4]){
+			case onechar('T', 0, 8):
+				return enum_loc_side::RIGHT;
+			break;
+			default: break;
+			}
+		break;
+		default: break;
+		}
+		break;
+	case 6:
+		switch(*((triehash_uu32*)&in[0])){
+		case onechar('B', 0, 32) | onechar('O', 8, 32) | onechar('T', 16, 32) | onechar('T', 24, 32):
+			switch(in[4]){
+			case onechar('O', 0, 8):
+				switch(in[5]){
+				case onechar('M', 0, 8):
+					return enum_loc_side::BOTTOM;
+				break;
+				default: break;
+				}
+			break;
+			default: break;
+			}
+		break;
+		default: break;
+		}
+		break;
+	default: break;
+	}
+	if(throw_on_invalid)
+		throw std::runtime_error("Found unrecognized enum value " + std::string(in) + " of enum_loc_side.");
+	return enum_loc_side::UXSD_INVALID;
+}
+
+inline enum_node_type lex_node_type(const char *in, bool throw_on_invalid){
```

Comment:
I'm wondering if something like this might work;
```cpp
/* Option 1 */
/* TOP|LEFT|RIGHT|BOTTOM */
inline enum_loc_side lex_loc_side(const char *in, bool throw_on_invalid){
	switch(in[0]){
	case 'T':
		if (strcmp("TOP\n", in))
			return enum_loc_side::TOP;
		goto lex_loc_side_error;
	case 'L':
		if (strcmp("LEFT\n", in))
			return enum_loc_side::LEFT;
		goto lex_loc_side_error;
	case 'R':
		if (strcmp("RIGHT\n", in))
			return enum_loc_side::RIGHT;
		goto lex_loc_side_error;
	case 'B':
		if (strcmp("BOTTOM\n", in))
			return enum_loc_side::BOTTOM;
		goto lex_loc_side_error;
	default:
		goto lex_loc_side_error;
	}
lex_loc_side_error:
	if(throw_on_invalid)
		throw std::runtime_error("Found unrecognized enum value " + std::string(in) + " of enum_loc_side.");
	return enum_loc_side::UXSD_INVALID;
}


/* Option 2 */
/* CHANX|CHANY|IPIN|OPIN|SOURCE|SINK */
inline enum_loc_side lex_loc_side(const char *in, bool throw_on_invalid){
	/* Compare first 4 bytes */
	switch(TO_INT32(in)){
	case TO_INT32("CHAN"):
		/* Compare next 2 bytes */
		switch(TO_INT16(in[4])) {
		case TO_INT16("X\0"):
			return enum_loc_side::CHANX;
		case TO_INT16("Y\0"):
			return enum_loc_side::CHANY;
		default:
			goto lex_loc_side_error;
		}
	case TO_INT32("IPIN"):
		/* Compare next byte */
		switch(TO_INT8(in[4])) {
		case TO_INT8('\0'):
			return enum_loc_side::IPIN;
		default:
			goto lex_loc_side_error;
		}
	case TO_INT32("OPIN"):
		/* Compare next byte */
		switch(TO_INT8(in[4])) {
		case TO_INT8('\0'):
			return enum_loc_side::OPIN;
		default:
			goto lex_loc_side_error;
		}
	case TO_INT32("SOUR"):
		/* Compare next 2 bytes */
		switch(TO_INT16(in[4])) {
		case TO_INT16("CE"):
			/* Compare next byte */
			switch(TO_INT8(in[6])) {
			case TO_INT8("\0"):
				return enum_loc_side::SOURCE;
			default:
				goto lex_loc_side_error;
			}

		default:
			goto lex_loc_side_error;
		}
	case TO_INT32("SINK"):
		/* Compare next byte */
		switch(TO_INT8(in[4])) {
		case TO_INT8('\0'):
			return enum_loc_side::SINK;
		default:
			goto lex_loc_side_error;
		}
	}
lex_loc_side_error:
	if(throw_on_invalid)
		throw std::runtime_error("Found unrecognized enum value " + std::string(in) + " of enum_loc_side.");
	return enum_loc_side::UXSD_INVALID;
}
```

Reply:
Might work, should test it.

At some point, I have to write byte-by-byte tries anyway, because some ARM processors are not very good at reading 4 or 8 bytes from an unaligned address: https://github.com/duck2/uxsdcxx/issues/21

Reply:
You might find [Google's RE2](https://github.com/google/re2/wiki/Syntax) can efficiently do this for you...

> RE2 is a fast, safe, thread-friendly alternative to backtracking regular expression engines like those used in PCRE, Perl, and Python. It is a C++ library.